keps/sig-scheduling/4671-gang-scheduling/README.md

```diff
@@ -0,0 +1,932 @@
+bogged down.
+-->
+
+#### Story 1: Gang-scheduling of a Job
+
+I have a tightly-coupled job and I want its pods to be scheduled and run only when the
+resources for all of them can be found in the cluster.
+
+#### Story 2: Gang-scheduling of a custom workload
+
+I have my own workload definition (CRD) and controller managing its lifecycle. I would
+like to be able to easily benefit of gang-scheduling feature supported by the core
+Kubernetes without extensive changes to my custom controller.
+
+
+### Risks and Mitigations
+
+#### The API needs to be extended in an unpredictable way
+
+We try to mitigate it by an extensive analysis of usecases and already sketching
+how we envision the direction in which the API will need to evolve to support further
+usecases. You can read more about it in the [extended proposal] document.
+
+
+## Design Details
+
+### Naming
+
+* `Workload` is the resource Kind.
+* `scheduling` is the ApiGroup.
+* `spec.workload` is the name of the new field in pod.
+* Within a Workload there is a list of groups of pods. Each group represents a top-level division of pods within a Workload.  Each group can be independently gang scheduled (or not use gang scheduling). This group is named
+  <<[UNRESOLVED community feedback requested]>> `PodGroup` or `GangGroup` for the top level. <<[/UNRESOLVED]>>.
+* In a future , we expect that this group can optionally specify further subdivision into sub groups.  Each sub-group can have an index.  The indexes go from 0 to N, without repeats or gaps. These subgroups are called
+  <<[UNRESOLVED depending on previous unresolved item]>> `PodSubGroup` if `PodGroup` is chosen, or else `RankedGroup` if `GangGroup` is chosen<<[/UNRESOLVED]>>.
+* In subsequent KEPs, we expect that a sub-group can optionally specify further subdivision into pod equivalence classes.  All pods in a pod equivalence class have the same values for all fields that affect scheduling feasibility.  These pod equivalence classes are called
+  <<[UNRESOLVED depending on a previous unresolved item]>> `PodSet` if `PodGroup` is chosen, or else `EqGroup` if `GangGroup` is chosen<<[/UNRESOLVED]>>.
+
+### Associating Pod into PodGroups
+
+When a `Workload` consists of a single group of pods needing Gang Scheduling, it is clear which pods belong to the group from the `spec.workload.name` field of the pod.  However `Workload` supports listing multiple list items, and a list item can represent a single group, or a set of identical replica groups.
+In these cases, there needs to be additional information to indicate which group a pod belongs to.
+
+We proposed to extend the newly introduced `pod.spec.workload` field with additional information
+to include that information. More specifically, the `pod.spec.workload` field is of type `PodWorkload`
+and is defined as following:
+
+```go
+type WorkloadReference struct {
+    // Workload defines the name of the Workload object this pod belongs to.
+    Workload string
```

Comment:
I prefer to rename it to either `workloadName` or just `name`. Also, should we mention that it is namespaced "same namespace as the Pod"?